File: java/BinarySearchStep.java
import java.io.*;
import java.util.*;

public class BinarySearchStep {
    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);
        try {
            // Prompt user for input file name
            System.out.print("Enter the CSV file name: ");
            String filename = scanner.nextLine();
            // Check if the filename is empty
            if (filename.isEmpty()) {
                System.out.println("No file name provided. Exiting.");
                return;
            }

            // Prompt user for target value
            System.out.print("Enter the target value to search for: ");
            int target;
            try {
                target = Integer.parseInt(scanner.nextLine());
                if (target < 0) {
                    System.out.println("Invalid target value. Exiting.");
                    return;
                }
            } catch (NumberFormatException e) {
                System.out.println("Invalid input for target value. Please enter a valid integer.");
                return;
            }

            List<Integer> numbers = new ArrayList<>();
            List<String> names = new ArrayList<>();
            // Read the CSV file and parse the numbers and names
            try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
                String line;
                while ((line = br.readLine()) != null) {
                    // Split the line by comma and parse the first part as an integer
                    // and the second part as a string (name)
                    String[] parts = line.split(",");
                    numbers.add(Integer.parseInt(parts[0].trim()));
                    names.add(parts[1].trim());
                }
            } catch (FileNotFoundException e) {
                System.out.println("File not found: " + filename);
                return;
            } catch (IOException e) {
                System.out.println("Error reading file: " + e.getMessage());
                return;
            }

            // Prepare output file to record each step of the binary search
            String outputFile = "binary_search_step_" + target + ".txt";
            PrintWriter out = new PrintWriter(outputFile);

            // Perform binary search and record each step
            int low = 0, high = numbers.size() - 1;
            while (low <= high) {
                int mid = (low + high) / 2;
                String step = mid + ": " + numbers.get(mid) + "/" + names.get(mid);
                // Output the current step to the file
                out.println(step);

                // Target found
                if (numbers.get(mid) == target) {
                    out.println("Target found at index: " + mid + " (" + names.get(mid) + ")");
                    out.close();
                    System.out.println("Target found at index: " + mid + " (" + names.get(mid) + ")");
                    System.out.println("Binary search steps recorded in: " + outputFile);
                    return;
                } else if (numbers.get(mid) < target) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }

            // Target not found, output -1
            out.println("-1");
            out.close();
            System.out.println("Target not found. Binary search steps recorded in: " + outputFile);
        } finally {
            scanner.close();
        }
    }
}

File: cpp/binary_search_step.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>
#include <sys/stat.h>

using namespace std;


int main() {
    try {
        string filename;
        int target;
        
        // Prompt user for the CSV file name
        cout << "Enter the CSV file name: ";
        getline(cin, filename);
        
        // Check if the filename is empty
        if (filename.empty()) {
            cout << "No file name provided. Exiting." << endl;
            return 0;
        }
        
        // Prompt user for the target value
        cout << "Enter the target value to search for: ";
        string targetInput;
        getline(cin, targetInput);
        
        try {
            target = stoi(targetInput);
            if (target < 0) {
                cout << "Invalid target value. Exiting." << endl;
                return 0;
            }
        } catch (const invalid_argument& e) {
            cout << "Invalid input for target value. Please enter a valid integer." << endl; 
            return 0;
        } catch (const out_of_range& e) {
            cout << "Target value out of range. Please enter a valid integer." << endl;
            return 0;
        }
        
        vector<int> numbers;
        vector<string> names;
        
        // Open the CSV file
        ifstream file(filename);
        if (!file.is_open()) {
            cout << "File not found: " << filename << endl; 
            return 0;
        }
        
        string line;
        // Read each line from the CSV file and parse number and name
        while (getline(file, line)) {
            stringstream ss(line);
            string numberStr, name;
            
            if (getline(ss, numberStr, ',') && getline(ss, name)) {
                // Trim whitespace from number and name
                numberStr.erase(0, numberStr.find_first_not_of(" \t"));
                numberStr.erase(numberStr.find_last_not_of(" \t") + 1);
                name.erase(0, name.find_first_not_of(" \t"));
                name.erase(name.find_last_not_of(" \t") + 1);
                
                // Convert number string to integer and store it with the name
                try {
                    numbers.push_back(stoi(numberStr)); 
                    names.push_back(name);
                } catch (const invalid_argument& e) {
                    cout << "Error parsing number in CSV: " << numberStr << endl;
                    return 0;
                }
            }
        }
        file.close();
        
        // Check if the numbers vector is empty
        if (numbers.empty()) {
            cout << "No data found in the file." << endl;
            return 0;
        }
        
        // Check if the numbers are sorted in non-decreasing order
        for (size_t i = 1; i < numbers.size(); i++) {
            if (numbers[i] < numbers[i - 1]) {
                cout << "The numbers in the file are not sorted. Please provide a sorted file." << endl;
                return 0;
            }
        }
        
        // Create output directory if it doesn't exist
        string outputFile = "binary_search_step_" + to_string(target) + ".txt";
        ofstream out(outputFile);
        
        if (!out.is_open()) {
            cout << "Error creating output file: " << outputFile << endl;
            return 0;
        }
        
        int low = 0, high = numbers.size() - 1;
        
        // Perform binary search and record each step
        while (low <= high) {
            int mid = (low + high) / 2;
            string step = to_string(mid) + ": " + to_string(numbers[mid]) + "/" + names[mid];
            
            // Output the current step
            out << step << endl;
            
            // Target found
            if (numbers[mid] == target) {
                out.close();
                cout << "Target found at index: " << mid << " (" << names[mid] << ")" << endl;
                cout << "Binary search steps recorded in: " << outputFile << endl;
                return 0;
            } else if (numbers[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        
        // Target not found, record -1
        out << "-1" << endl;
        out.close();
        cout << "Target not found. Binary search steps recorded in: " << outputFile << endl;
    

    } catch (const exception& e) {
        cout << "An error occurred: " << e.what() << endl;
        return 1;
    }
    return 0;
}